javascript/ql/src/semmle/javascript/GlobalAccessPaths.qll

```diff
@@ -415,7 +415,7 @@ module AccessPath {
   pragma[inline]
   DataFlow::SourceNode getAnAliasedSourceNode(DataFlow::Node node) {
     exists(DataFlow::SourceNode root, string accessPath |
-      node = AccessPath::getAReferenceTo(root, accessPath) and
+      node = pragma[only_bind_into](AccessPath::getAReferenceTo(root, accessPath)) and
```

Comment:
> This looks backwards to me. Which call site was this needed for?

This call site: 
https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/DynamicPropertyAccess.qll#L44-L46

I tried to put the pragma in 4 different places (lhs and rhs of the two equalities). 
And I only got a good join-order with the one in the PR. 

Here is the resulting join-order with the pragma I'm suggesting: 
```
150    ~0%      {2} r1 = SCAN DynamicPropertyAccess::EnumeratedPropName::getSourceObject_dispred#ff OUTPUT In.1, In.0 'this'
164    ~3%      {2} r2 = JOIN r1 WITH Sources::Cached::hasLocalSource#ff@staged_ext ON FIRST 1 OUTPUT Rhs.1 'result', Lhs.1 'this'
164    ~4%      {2} r3 = JOIN r2 WITH Sources::SourceNode::Range#class#f@staged_ext ON FIRST 1 OUTPUT Lhs.1 'this', Lhs.0 'result'
131    ~0%      {3} r4 = JOIN r1 WITH GlobalAccessPaths::AccessPath::getAReferenceTo#2#fff_201#join_rhs ON FIRST 1 OUTPUT Rhs.1, Rhs.2, Lhs.1 'this'
652    ~0%      {2} r5 = JOIN r4 WITH GlobalAccessPaths::AccessPath::getAReferenceTo#2#fff ON FIRST 2 OUTPUT Rhs.2 'result', Lhs.2 'this'
166    ~3%      {2} r6 = JOIN r5 WITH Sources::SourceNode::Range#class#f@staged_ext ON FIRST 1 OUTPUT Lhs.1 'this', Lhs.0 'result'
330    ~72%     {2} r7 = r3 UNION r6
                return r7
```

My rough understanding is that the compiler cannot use `getAReferenceTo` to restrict `node`, it can only use `node` to restrict `getAReferenceTo`.  
It therefore has to somehow compute `node` first, and the obvious way to do that is to specialize `node` in the predicate that `getAnAliasedSourceNode` gets inlined into.   
I might be wrong though.

Reply:
> My rough understanding is that the compiler cannot use `getAReferenceTo` to restrict `node`, it can only use `node` to restrict `getAReferenceTo`.
> It therefore has to somehow compute `node` first, and the obvious way to do that is to specialize `node` in the predicate that `getAnAliasedSourceNode` gets inlined into.

This is correct and exactly the reason why this works. `getSourceObject` restricts `node`, which then by the pragma _binds into_ the result column of `getAReferenceTo` and everything is good and the join-orderer sees this as a reasonable choice. If the join-orderer instead attempts to put `getAReferenceTo` first then it can't select `getSourceObject` without making things look like a cartesian product (node isn't bound), so that's why it now won't select `getAReferenceTo` first.

Reply:
After a chat with @aschackmull, everything looks good. 👍 

It seems I desugared the form `x = pragma(foo(a,b))` incorrectly. I thought of it as `foo(a, b, pragma(x))`, whereas the correct desugaring is `x = pragma(tmp) and foo(a, b, tmp)`.